packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -491,7 +491,11 @@ export class HardhatNode extends EventEmitter {
     blockNumberOrPending: bigint | "pending",
     stateOverrideSet: StateOverrideSet = {}
   ): Promise<RunCallResult> {
-    const tx = await this._getTransactionForCall(call, blockNumberOrPending);
+    const tx = await this._getTransactionForCall(
+      call,
+      blockNumberOrPending,
+      stateOverrideSet
+    );
```

Comment:
I'm curious why you added this, because I'm not really sure it's necessary. The description of state overrides is a bit ambiguous about this.

Reply:
To expand on this:

I don't know if `eth_call({from: a, nonce: X})` is the same as `eth_call({from: a}, {[a]: { nonce: X }})`.

My understanding here is that state overrides mean "make these changes before the execution", so in most scenarios these would be equivalent. The only scenario where _maybe_ there's a difference is during deployments. But observing differences in `eth_call`s that deploy something is really hard. And to make everything worse, geth seems to completely ignore the nonces for `eth_call` "deployments". Which might be a bug or not, I don't know.

The bottom line for me here is: if we are not really sure that this is necessary, I would drop it. This is not a very specified issue, so if we have to guess, let's guess whatever results in simpler code.

Reply:
AFAIK, the nonce is extracted from the state. When we are running the transaction in `revm` it validates that the nonce of the transaction is not less than the nonce of the state, throwing an error.

Thus, if we don't use the overriden nonce when determining the transaction nonce, the transaction is invalid, I think.

Reply:
I see, in that case it makes sense to do it :+1: